### PR TITLE
Missing address for provider

### PR DESCRIPTION
**Pull request by defong**
### Context
Lack of ordering of provider enrichment where by older version of enrichment details are used instead of new version.

The address section of the enrichment can be empty/blank, ie only enriched the section concerning their accrediting provider, the blanked values are used.

### Changes proposed in this pull request
Added ordering of provider enrichment based on created_at date as default
Fixed the address to hydrate based on using provider details if enrichment is not available

### CODE REVIEW

File: app/models/provider.rb [hunk 1]
```diff
@@ -42,8 +42,23 @@ class Provider < ApplicationRecord
   has_many :sites
   has_many :enrichments, foreign_key: :provider_code, primary_key: :provider_code, class_name: "ProviderEnrichment"
   def address_info
-    (enrichments.with_address_info.last || self)
+    fields = %w[address1 address2 address3 address4 postcode]
+
+
+    result = enrichments.with_address_info.first.present? && enrichments.with_address_info.first
       .attributes_before_type_cast
-      .slice('address1', 'address2', 'address3', 'address4', 'postcode', 'region_code')
```
Comment (tvararu) [inline app/models/provider.rb]: No more need for `region_code`?

Comment (tvararu) [reply app/models/provider.rb]: Oh it's lower down

File: app/models/provider.rb [hunk 2]
```diff
@@ -42,8 +42,23 @@ class Provider < ApplicationRecord
   has_many :sites
   has_many :enrichments, foreign_key: :provider_code, primary_key: :provider_code, class_name: "ProviderEnrichment"
   def address_info
-    (enrichments.with_address_info.last || self)
+    fields = %w[address1 address2 address3 address4 postcode]
+
+
+    result = enrichments.with_address_info.first.present? && enrichments.with_address_info.first
```
Comment (benilovj) [inline app/models/provider.rb]: I find this a bit hard to parse...

File: app/models/provider.rb [hunk 3]
```diff
@@ -42,8 +42,29 @@ class Provider < ApplicationRecord
...
+    # region_code should be present in enrichment regardless
+    result['region_code'] = (enrichments.with_address_info.first || self).region_code_before_type_cast
+
+    result
+  end
+
+  def valid_enriched_address_info
+    fields = %w[address1 address2 address3 address4 postcode]
+
+    result = enrichments.with_address_info.first.present? && enrichments.with_address_info.first
```
Comment (benilovj) [inline app/models/provider.rb]: I think you can clean this up even further by using a functional style:

```ruby
return nil unless enrichments.with_address_info.present?

results = enrichments.with_address_info.first
                ...
```

Comment (benilovj) [reply app/models/provider.rb]: Alternatively:

```ruby
def valid_enriched_address_info
  fields = %w[address1 address2 address3 address4 postcode]

  return nil unless enrichments.with_address_info.present?

  enrichments.with_address_info.first
    .attributes_before_type_cast
    .slice(*fields)
    .select {|k,v| v.present? }
end
```

Comment (defong) [reply app/models/provider.rb]: cannot use select directly as its omits the field in question

**Comment by defong** [comment]
[Missing addresses in the Providers endpoint of the UCAS API](https://trello.com/c/yyxhjI7k/901-missing-addresses-in-the-providers-endpoint-of-the-ucas-api)

**Comment by misaka** [comment]
Sorry about the CircleCI failure, didn't realise it'd be inserted as a requirement for the PR.

**Comment by misaka** [comment]
Ah, ok, it's not blocking, it just looks angry. I'll disable the integration for now.

File: app/models/provider.rb [hunk 4]
```diff
@@ -42,8 +42,30 @@ class Provider < ApplicationRecord
...
+    fields = %w[address1 address2 address3 address4 postcode]
+
+    result = enrichments.with_address_info.first
+     .attributes_before_type_cast
+     .slice(*fields)
+
+    return nil unless (result.select { |_k, v| v.present? }).present?
+
+    result['region_code'] = enrichments.with_address_info.first.region_code_before_type_cast
+    result
```
[the remark this answers is not in the document]
Comment (defong) [reply app/models/provider.rb]:
```
.json_data_where_not(address1: nil, address2: nil, address3: nil, address4: nil, postcode: nil)
``` used that instead as region_code will cause a false positive

File: spec/models/provider_spec.rb [hunk 5]
```diff
@@ -34,38 +34,146 @@
...
+              'address1' => provider.address1,
+              'address2' => provider.address2,
+              'address3' => provider.address3,
+              'address4' => provider.address4,
+              'postcode' => provider.postcode,
+              'region_code' => ProviderEnrichment.region_codes[provider.region_code]
+            )
+          end
+          it 'returns address of the provider' do
+            enrichment = build(:provider_enrichment, json_data: { 'email' => Faker::Internet.email,
```
Comment (misaka) [inline spec/models/provider_spec.rb]: Set the attributes you want here, `address1`, `address2`, etc as object attributes, not as `json_data` attributes;

File: app/models/provider.rb [hunk 6]
```diff
@@ -62,8 +62,10 @@ class Provider < ApplicationRecord
 
   # TODO: filter to published enrichments, maybe rename to published_address_info
   def address_info
-    (enrichments.with_address_info.last || self)
+    fields = %w[address1 address2 address3 address4 postcode region_code]
+
+    (enrichments.latest_created_at.with_address_info.first || self)
       .attributes_before_type_cast
-      .slice('address1', 'address2', 'address3', 'address4', 'postcode', 'region_code')
+      .slice(*fields)
```
Comment (misaka) [inline app/models/provider.rb]: Just an opinion, but I think that `fields` is short enough that it can be inline here without getting in the way or readability.

File: spec/models/provider_spec.rb [hunk 7]
```diff
@@ -34,38 +34,146 @@
...
+              'train_with_disability' => Faker::Lorem.sentence.to_s })
+              provider = create(:provider, region_code: london, enrichments: [enrichment])
+
+              expect(provider.address_info).to eq(
+                'address1' => provider.address1,
+                'address2' => provider.address2,
+                'address3' => provider.address3,
+                'address4' => provider.address4,
+                'postcode' => provider.postcode,
+                'region_code' => ProviderEnrichment.region_codes[provider.region_code]
```
Comment (misaka) [inline spec/models/provider_spec.rb]: Change `ProviderEnrichment.region_codes[provider.region_code]` to `provider.region_code_before_type_cast`.

File: spec/models/provider_spec.rb [hunk 8]
```diff
@@ -34,38 +34,146 @@
...
+
+          context 'enrichment has only region code set for address_info' do
+            it 'returns address of the provider' do
+              london = ProviderEnrichment.region_codes['London']
+              enrichment = build(:provider_enrichment, json_data: { 'email' => Faker::Internet.email,
+              'website' => Faker::Internet.url,
+              'region_code' => 'No region',
+              'train_with_us' => Faker::Lorem.sentence.to_s,
+              'train_with_disability' => Faker::Lorem.sentence.to_s })
+              provider = create(:provider, region_code: london, enrichments: [enrichment])
```
Comment (misaka) [inline spec/models/provider_spec.rb]: We need to remove all the address-related fields other than `region_code` from the actual JSON blob to properly simulate what `manage-courses-api` stuffs into the DB. So use this bit of code from above:

```ruby
            # forcing it to be absent in json_data altogether
            ProviderEnrichment.connection.update(<<~EOSQL)
              UPDATE provider_enrichment
                    SET json_data=json_data-'Address1'-'Address2'-'Address3'-'Address4'-'Postcode'-'RegionCode'
                    WHERE provider_code='#{enrichment.id}'
            EOSQL
```

File: app/models/provider_enrichment.rb [hunk 9]
```diff
@@ -24,7 +25,10 @@ class ProviderEnrichment < ApplicationRecord
   belongs_to :provider, foreign_key: :provider_code, primary_key: :provider_code
 
   scope :with_address_info,
-        -> { where("json_data ?| array['Address1', 'Address2', 'Address3', 'Address4', 'Postcode', 'RegionCode']") }
+        -> do
+          where("json_data ?| array['Address1', 'Address2', 'Address3', 'Address4', 'Postcode', 'RegionCode']")
```
Comment (misaka) [inline app/models/provider_enrichment.rb]: We probably need to remove `RegionCode` from here.

File: app/models/provider.rb (hunk 2 above)
Comment (defong) [reply app/models/provider.rb]: staled

File: app/models/provider.rb (hunk 3 above)
Comment (defong) [reply app/models/provider.rb]: staled

File: spec/models/provider_spec.rb [hunk 10]
```diff
@@ -142,27 +137,54 @@
...
+            end
+
             it 'returns address of the provider' do
-              london = ProviderEnrichment.region_codes['London']
-              enrichment = build(:provider_enrichment, json_data: { 'email' => Faker::Internet.email,
-              'website' => Faker::Internet.url,
-              'region_code' => 'No region',
-              'train_with_us' => Faker::Lorem.sentence.to_s,
-              'train_with_disability' => Faker::Lorem.sentence.to_s })
+              enrichment = build(:provider_enrichment, region_code: no_region, address2: nil,
```
Comment (misaka) [inline spec/models/provider_spec.rb]: How about a new context for this to be clear that we test the way the API works?

File: app/models/provider_enrichment.rb [hunk 11]
```diff
@@ -16,6 +16,7 @@
 class ProviderEnrichment < ApplicationRecord
   self.table_name = "provider_enrichment"
   self.primary_key = "provider_code"
+  scope :latest_created_at, -> { order(created_at: :desc) }
```
Comment (dankmitchell) [inline app/models/provider_enrichment.rb]: Can we move this done to line 32 so all the "scopes" are grouped together

File: spec/models/provider_spec.rb [hunk 12]
```diff
@@ -70,19 +67,18 @@
...
-            enrichment = build(:provider_enrichment, json_data: { 'email' => Faker::Internet.email,
-            'website' => Faker::Internet.url,
-            'train_with_us' => Faker::Lorem.sentence.to_s,
-            'train_with_disability' => Faker::Lorem.sentence.to_s,
-            'address1' => nil,
-            'address2' => nil,
-            'address3' => nil,
-            'address4' => nil,
-            'postcode' => nil })
+            enrichment = build(:provider_enrichment,
```
Comment (misaka) [inline spec/models/provider_spec.rb]: How about a new context for this to be clear that we test the way the API works?

**Comment by misaka** [comment]
@defong Looking good, thanks. Last couple of small tidbits.

File: spec/requests/api/v1/provider_spec.rb [hunk 13]
```diff
@@ -84,7 +84,7 @@
               headers: { 'HTTP_AUTHORIZATION' => credentials }
 
           json = JSON.parse(response.body)
-          expect(json). to eq(
+          expect(json). to match_array(
```
Comment (misaka) [inline spec/requests/api/v1/provider_spec.rb]: 👍

Comment (defong) [reply spec/requests/api/v1/provider_spec.rb]: thanks that worked as a treat
